interpreter/cling/lib/Interpreter/ValuePrinter.cpp

```diff
@@ -344,8 +344,452 @@ static void StreamValue(llvm::raw_ostream& o, const void* V,
+    std::ostringstream strm;
+    strm << val;
+    return strm.str();
+  }
+
+  std::string printValue(unsigned long long val) {
+    std::ostringstream strm;
+    strm << val;
+    return strm.str();
+  }
+
+  std::string printValue(float val) {
+    std::ostringstream strm;
+    strm << val;
+    return strm.str();
+  }
+
+  std::string printValue(double val) {
+    std::ostringstream strm;
+    strm << val;
+    return strm.str();
+  }
+
+  std::string printValue(long double val) {
+    std::ostringstream strm;
+    strm << val << "L";
+    return strm.str();
+  }
+
+  // Char pointers
+  std::string printValue(const char *const val) {
+    std::ostringstream strm;
+    if (!val) {
+      strm << "<<<NULL>>>";
+    } else {
+      const char *cobj = val;
+      int i = 0;
+      strm << "\"";
+      while (*cobj != 0) {
+        strm << *cobj;
+        cobj++;
+        i++;
+      }
+      strm << "\"";
+    }
+    return strm.str();
+  }
+
+  std::string printValue(char *val) {
+    return printValue((const char *const) val);
+  }
+
+  // std::string
+  std::string printValue(const std::string & val) {
+    std::ostringstream strm;
+    strm << "\"";
+    strm << val;
+    strm << "\"";
+    return strm.str();
+  }
+
+  std::string printValue(const Value &value) {
+    std::ostringstream strm;
+
+    if (!value.isValid()) {
+      strm << "<<<invalid>>> @" << &value;
+    } else {
+      clang::ASTContext& C = value.getASTContext();
+      clang::QualType QT = value.getType();
+      strm << "boxes [";
+      strm << "("
+      << QT.getAsString(C.getPrintingPolicy())
+      << ") ";
+      clang::QualType valType = QT.getDesugaredType(C).getNonReferenceType();
+      if (C.hasSameType(valType, C.LongDoubleTy))
+        strm << value.simplisticCastAs<long double>() << "L";
+      else if (valType->isFloatingType())
+        strm << value.simplisticCastAs<double>();
+      else if (valType->isIntegerType()) {
+        if (valType->hasSignedIntegerRepresentation())
+          strm << value.simplisticCastAs<long long>();
+        else
+        strm << value.simplisticCastAs<unsigned long long>();
+      } else if (valType->isBooleanType())
+        strm << (value.simplisticCastAs<bool>() ? "true" : "false");
+      else if (!valType->isVoidType()) {
+        strm << invokePrintValueOverload(value);
+      }
+      strm << "]";
+    }
+
+    return strm.str();
+  }
+
 namespace valuePrinterInternal {
+
+  std::string printType_New(const Value& V) {
+    using namespace clang;
+    std::ostringstream strm;
+    QualType QT = V.getType().getNonReferenceType();
+    std::string ValueTyStr;
+    if (const TypedefType* TDTy = dyn_cast<TypedefType>(QT))
+      ValueTyStr = TDTy->getDecl()->getQualifiedNameAsString();
+    else if (const TagType* TTy = dyn_cast<TagType>(QT))
+      ValueTyStr = TTy->getDecl()->getQualifiedNameAsString();
+
+    if (ValueTyStr.empty())
+      ValueTyStr = QT.getAsString();
+    else if (QT.hasQualifiers())
```

Comment:
I don't understand that `else if`. Why do we only print the qualifiers if we have a `ValueTyStr`?